rust/opendp/src/trans/count.rs

```diff
@@ -4,77 +4,56 @@ use std::convert::TryFrom;
+use crate::dist::{AbsoluteDistance, SymmetricDistance, LpDistance};
 use crate::dom::{AllDomain, MapDomain, SizedDomain, VectorDomain};
 use crate::error::*;
 use crate::traits::DistanceConstant;
 
-pub fn make_count<MI, TI, TO>() -> Fallible<Transformation<VectorDomain<AllDomain<TI>>, AllDomain<TO>, MI, AbsoluteDistance<TO>>>
-    where MI: DatasetMetric,
-          TO: TryFrom<usize> + Bounded + One + DistanceConstant {
+pub fn make_count<TIA, TO>(
+) -> Fallible<Transformation<VectorDomain<AllDomain<TIA>>, AllDomain<TO>, SymmetricDistance, AbsoluteDistance<TO>>>
+    where TO: TryFrom<usize> + Bounded + One + DistanceConstant {
     Ok(Transformation::new(
         VectorDomain::new_all(),
         AllDomain::new(),
-        // min(arg.len(), u32::MAX)
-        Function::new(move |arg: &Vec<TI>| TO::try_from(arg.len()).unwrap_or(TO::max_value())),
-        MI::default(),
+        // think of this as: min(arg.len(), TO::max_value())
+        Function::new(move |arg: &Vec<TIA>| TO::try_from(arg.len()).unwrap_or(TO::max_value())),
+        SymmetricDistance::default(),
         AbsoluteDistance::default(),
         StabilityRelation::new_from_constant(TO::one())))
 }
 
 
-pub fn make_count_distinct<MI, MO, TI>() -> Fallible<Transformation<VectorDomain<AllDomain<TI>>, AllDomain<MO::Distance>, MI, MO>>
-    where MI: DatasetMetric,
-          MO: SensitivityMetric,
-          MO::Distance: TryFrom<usize> + Bounded + One + DistanceConstant,
-          TI: Eq + Hash {
+pub fn make_count_distinct<TIA, TO>() -> Fallible<Transformation<VectorDomain<AllDomain<TIA>>, AllDomain<TO>, SymmetricDistance, AbsoluteDistance<TO>>>
+    where TIA: Eq + Hash,
+          TO: TryFrom<usize> + Bounded + One + DistanceConstant {
     Ok(Transformation::new(
         VectorDomain::new_all(),
         AllDomain::new(),
-        Function::new(move |arg: &Vec<TI>| {
+        Function::new(move |arg: &Vec<TIA>| {
             let len = arg.iter().collect::<HashSet<_>>().len();
-            MO::Distance::try_from(len).unwrap_or(MO::Distance::max_value())
+            TO::try_from(len).unwrap_or(TO::max_value())
         }),
-        MI::default(),
-        MO::default(),
-        StabilityRelation::new_from_constant(MO::Distance::one())))
-}
-
-pub trait CountByConstant<MI: DatasetMetric, MO: SensitivityMetric> {
-    fn get_stability_constant() -> Fallible<MO::Distance>;
-}
-
-impl<QO: NumCast> CountByConstant<HammingDistance, L1Distance<QO>> for (HammingDistance, L1Distance<QO>) {
-    fn get_stability_constant() -> Fallible<QO> {
-        num_cast!(2.; QO)
-    }
+        SymmetricDistance::default(),
+        AbsoluteDistance::default(),
+        StabilityRelation::new_from_constant(TO::one())))
 }
 
-impl<QO: FloatConst> CountByConstant<HammingDistance, L2Distance<QO>> for (HammingDistance, L2Distance<QO>) {
-    fn get_stability_constant() -> Fallible<QO> {
-        Ok(QO::SQRT_2())
-    }
+pub trait CountByConstant: SensitivityMetric {
```

Comment:
It feels weird to make this a sub-trait of `SensitivityMetric`. I get the desire to reduce the number of trait bounds in function signatures, but I think this one actually makes the bounds on `make_count_by_categories()` a little cryptic. I think it'd be more clear with an explicit `where MO: SensitivityMetric + CountByConstant`.

Reply:
Hmm, the motivation wasn't just to reduce the number of bounds- the stability constant is only well defined with respect to a metric, and it returns a constant of the distance type of that metric. That being said, this still works:

```rust
pub trait CountByConstant<QO> {
    fn get_stability_constant() -> QO;
}
impl<Q: One, const P: usize> CountByConstant<Q> for LpDistance<Q, P> {
    fn get_stability_constant() -> Q { Q::one() }
}
```

With updated trait bound:
`where MO: CountByConstant<MO::Distance> + SensitivityMetric,`

I'm fine with either.

Reply:
Ah OK, makes sense, thanks!